### ISSUE TITLE
Redshift: SIMILAR TO pattern matching operators are not parsed

### ISSUE THREAD

**Issue by PLBMR** [sqlfluff/sqlfluff#2722]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

I ran `sqlfluff lint --dialect=redshift --ignore templating example.sql` with the `.sqlfluff` specified below and `example.sql` with the following contents:

```sql
WITH cleaned_bear_financial_branch AS (
    SELECT
        branch_id,
        TO_NUMBER(CASE WHEN honey_numerical_code SIMILAR TO '[0-9]{0,7}.?[0-9]{0,2}' THEN honey_numerical_code ELSE NULL END, '24601') AS honey_numerical_code
    FROM bear_financial_branch
)

SELECT branch_id
FROM cleaned_bear_financial_branch
LIMIT 10;
```

I got this error:

```bash
== [example.sql] FAIL                                                                                                                                                             
L:   4 | P:  19 |  PRS | Line 4, Position 19: Found unparsable section: 'CASE WHEN honey_numerical_code SIMILAR T...'
All Finished 📜 🎉!
```

From some investigations with `sqlfluff parse --ignore templating example.sql`, it seems apparent that the `redshift` dialect does not currently support [`SIMILAR TO` pattern matching operators](https://docs.aws.amazon.com/redshift/latest/dg/pattern-matching-conditions-similar-to.html). This dialect should be supporting `SIMILAR TO` operators.

### Expected Behaviour

We expect `example.sql` as specified above to have no parse errors.

### Observed Behaviour

We observe that `example.sql` as specified above has the following parse error:

```bash
== [example.sql] FAIL                                                                                                                                                             
L:   4 | P:  19 |  PRS | Line 4, Position 19: Found unparsable section: 'CASE WHEN honey_numerical_code SIMILAR T...'
All Finished 📜 🎉!
```

### How to reproduce

1. Have the `.sqlfluff` file as specified below in your working directory.
2. Install requirements: `pip install dbt==0.20.0 sqlfluff==0.10.1 markupsafe==2.0.1`
3. Have `example.sql` as specified above in your working directory.
4. Run `sqlfluff lint --dialect=redshift --ignore templating example.sql`.

### Dialect

`redshift`

### Version

`sqlfluff==0.10.1`

### Configuration

- `.sqlfluff`:
```
[sqlfluff]
verbose = 0
nocolor = False
dialect = redshift
templater = jinja
rules = None
exclude_rules = L009,L011,L019,L029,L031,L034,L054
recurse = 0
output_line_length = 160
runaway_limit = 10
ignore_templated_areas = True
encoding = autodetect
# Comma separated list of file extensions to lint.

# NB: This config will only apply in the root folder.
sql_file_exts = .sql,.sql.j2,.dml,.ddl

[sqlfluff:indentation]
indented_joins = False
indented_using_on = True
template_blocks_indent = True

[sqlfluff:templater]
unwrap_wrapped_queries = True

[sqlfluff:templater:jinja]
apply_dbt_builtins = True

[sqlfluff:templater:jinja:macros]
# Macros provided as builtins for dbt projects
dbt_ref = {% macro ref(model_ref) %}{{model_ref}}{% endmacro %}
dbt_source = {% macro source(source_name, table) %}{{source_name}}_{{table}}{% endmacro %}
dbt_config = {% macro config() %}{% for k in kwargs %}{% endfor %}{% endmacro %}
dbt_var = {% macro var(variable) %}item{% endmacro %}
dbt_is_incremental = {% macro is_incremental() %}True{% endmacro %}

# Some rules can be configured directly from the config common to other rules.
[sqlfluff:rules]
tab_space_size = 4
max_line_length = 160
indent_unit = space
comma_style = trailing
allow_scalar = True
single_table_references = consistent
unquoted_identifiers_policy = all

# Some rules have their own specific config.
[sqlfluff:rules:L003]
lint_templated_tokens = True

[sqlfluff:rules:L007]  # Keywords
operator_new_lines = after

[sqlfluff:rules:L010]  # Keywords
capitalisation_policy = consistent

[sqlfluff:rules:L014]  # Unquoted identifiers
extended_capitalisation_policy = consistent

[sqlfluff:rules:L016]
ignore_comment_lines = False

[sqlfluff:rules:L029]  # Keyword identifiers
unquoted_identifiers_policy = aliases

[sqlfluff:rules:L030]  # Function names
capitalisation_policy = consistent

[sqlfluff:rules:L038]
select_clause_trailing_comma = forbid

[sqlfluff:rules:L040]  # Null & Boolean Literals
capitalisation_policy = consistent

[sqlfluff:rules:L042]
# By default, allow subqueries in from clauses, but not join clauses.
forbid_subquery_in = join

[sqlfluff:rules:L047]  # Consistent syntax to count all rows
prefer_count_1 = False
prefer_count_0 = False
```

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by PLBMR** [comment]
Update: These should also be supported in the `postgres` dialect (see [9.7.2](https://www.postgresql.org/docs/14/functions-matching.html)). We see this is not currently supported in our `postgres` dialect either:

```bash
sqlfluff lint --dialect=postgres --ignore templating bug_reproductions/analytics_dbt_issue_01/ex
ample.sql 
== [bug_reproductions/analytics_dbt_issue_01/example.sql] FAIL                                                                                                                    
L:   4 | P:  19 |  PRS | Line 4, Position 19: Found unparsable section: 'CASE
                       | WHEN honey_numerical_code SIMILAR T...'
```

Will make my change so that `SIMILAR TO` is supported by both the `postgres` dialect and the `redshift` dialect.

**Comment by PLBMR** [comment]
Update: These should also be supported in the `postgres` dialect (see [9.7.2](https://www.postgresql.org/docs/14/functions-matching.html)). We see this is not currently supported in our `postgres` dialect either:

```bash
>>> sqlfluff lint --dialect=postgres --ignore templating example.sql 
== [bug_reproductions/analytics_dbt_issue_01/example.sql] FAIL                                                                                                                    
L:   4 | P:  19 |  PRS | Line 4, Position 19: Found unparsable section: 'CASE
                       | WHEN honey_numerical_code SIMILAR T...'
```

Will make my change so that `SIMILAR TO` is supported by both the `postgres` dialect and the `redshift` dialect.

**Comment by PLBMR** [comment]
Testing out the change a bit. While `ansi` does support these pattern matching operators, there are unintended downstream side effects on the `tsql` dialect when it inherits these changes from `ansi`. Making adjustments to `tsql` to prevent these side effects will bloat the MR for this issue.

My plan:
1. To address this issue, I will only allow `postgres` and `redshift` dialects to support the `SIMILAR TO` operator.
2. After creating an MR for this issue, I will create an issue to allow the `ansi` dialect to support the `SIMILAR TO` operator, without creating side effects for descendent dialects.

**Comment by PLBMR** [comment]
To make this MR relatively simple, I broke out some followups for the `ansi`, `postgres`, and `redshift` dialects in #2730 and #2731 .
